specs/makeStyles.md

```diff
@@ -0,0 +1,290 @@
+};
+```
+
+This object uses CSS property names as keys, builds "compound" string keys for nested objects to flatten the output. With that approach, it should be quite cheap to apply overrides. It is not necessary to read the object values, we just need to shallow apply (merge) the override object on top of the base object. No style processing is necessary:
+
+```js
+merged = {
+  ...baseClasses,
+  ...overrideClasses,
+```
+
+After the merge, we have all the classes which need to be applied to the element being styled.
+
+### Mapping props/state to styles
+
+The output styles of a component change based on its props and state.
+For that the `makeStyles` call accepts an array of items where each item is a `[matcher; styles]` pair:
+
+```js
+useStyles = makeStyles([
+	[matcher1, styles1],
+	[matcher2, styles2],
+  ...
+]);
+```
+
+Each `matcher` is a function with takes a consumer-defined input and returns a boolean.
+When `useStyles` is called it calls all the matchers from top to bottom, passing its first argument to them. For every matcher which returns truthy, its styles are used.
+
+There is a special `null` matcher which always matches. We might want to change the always-matching signature to `() => true` to get rid of the special value or use a more descriptive constant like `MATCH_ALWAYS`.
+
+### Applying theme to styles
+
+No matter what theme is used, the component styles are always the same. The only way how to change the component appearance is through tokens which can be used in style values.
+The `styles` passed to `makeStyles([[matcher, styles]])` call can be either an object or a function. The function is called with all theme tokens.
+Those tokens are resolved to CSS variable usages in build time. `ThemeProvider` component is responsible for setting the CSS variables in DOM and changing them when theme changes. When theme is switched, only the variables are changed, all styles remain the same.
+
+::Risk:: What if we will not be able to describe the theme changes using just different values?
+::Risk:: A theme consist of hundreds or even thousands of tokens. We need to verify browsers can handle this amount of CSS variables.
+
+#### IE11 compatibility
+
+IE11 does not support CSS variables. If we decide to support IE11, we will split the styles to the items which use tokens and those which do not. The ones which do not use tokens will be processed build time, for the ones with tokens, we will fall back to runtime evaluation.
+
+### Hash-based classnames
+
+We will use a hash (MurmurHash2) of the CSS property name and the value to generate a classname. That results in deterministic (=stable) classnames. With that we can support server side rendering.
+
+### Passing overrides to children using `className`
+
+To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary sideeffect described in `useCSS` approach above.
```

Comment:
```suggestion
To avoid the ambivalence of using `className` vs `styles` props when passing overrides to children based on their type, we will use the dictionary side effect described in `useCSS` approach above.
```